File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.dfs(root)[1]

    def dfs(self, root):
        if not root:
            return [0, True]

        left = self.dfs(root.left)
        right = self.dfs(root.right)

        balanced = left[1] and right[1] and abs(left[0] - right[0]) <= 1

        return [1 + max(left[0], right[0]), balanced]

    
    
        


            

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) {
            let res = 0;

            function dfs(root) {
                if (!root) return 0; 

                const leftDepth = dfs(root.left);
                const rightDepth = dfs(root.right);

                res = Math.max(res, leftDepth + rightDepth);

                return 1 + Math.max(leftDepth, rightDepth);
            }

            dfs(root);

            return res;
    }
}
